File: apps/ios/Voiid/Voiid/Games/Carrom/CarromGame.swift
//
//  CarromGame.swift
//  Voiid Ui
//
//  Main observable game controller for Carrom.
//  Coordinates physics steps, turn lifecycle, gesture handling, rules evaluation,
//  haptic feedback, and bot execution.
//

import SwiftUI
import Combine

enum CarromPhase: Equatable {
    case placement
    case aiming
    case inMotion
    case evaluating
    case gameOver(winner: String)
}

enum CarromBannerType {
    case info, success, foul
}

struct CarromBanner: Equatable {
    let text: String
    let type: CarromBannerType
}

@MainActor
final class CarromGame: ObservableObject {

    // MARK: - Published State

    @Published var mode: CarromGameMode = .classic
    @Published var isBot: Bool = true
    @Published var turn: CarromTurn = .player1
    @Published var phase: CarromPhase = .placement

    @Published var player1Score: Int = 0
    @Published var player2Score: Int = 0
    @Published var player1Pieces: [CarromPieceType] = []
    @Published var player2Pieces: [CarromPieceType] = []

    @Published var baselineNormX: CGFloat = 0.5
    @Published var aimDirection: CGVector = CGVector(dx: 0, dy: -1) // Shooting UP by default
    @Published var shotPower: CGFloat = 0.0 // 0.0 ... 1.0
    @Published var isDraggingAim: Bool = false
    @Published var isPlacementValid: Bool = true

    @Published var aimResult: CarromAimResult? = nil
    @Published var banner: CarromBanner? = nil
    @Published private(set) var animationTick: UInt = 0

    var canControl: Bool {
        (phase == .placement || phase == .aiming) && (turn == .player1 || !isBot)
    }

    private var pendingAction: (delay: TimeInterval, action: () -> Void)?

    // Delays advance only while the screen is active. Resetting drops old bot/turn work.
    private func schedule(after delay: TimeInterval, action: @escaping () -> Void) {
        pendingAction = (delay, action)
    }

    func pauseClock() { lastFrameTime = 0 }

    // Queen Tracking
    @Published var queenState: QueenState = .onBoard
    enum QueenState: Equatable {
        case onBoard
        case pendingCover(by: CarromTurn)
        case covered(by: CarromTurn)
    }

    // Win condition
    let targetScore: Int = 9

    // MARK: - Internal Physics Engine

    let world = CarromPhysicsWorld()
    private var pocketedThisTurn: [CarromPiece] = []
    private var lastFrameTime: TimeInterval = 0

    init(mode: CarromGameMode = .classic, isBot: Bool = true) {
        self.mode = mode
        self.isBot = isBot
        setupWorldCallbacks()
        resetBoard()
    }

    // MARK: - World Setup & Callbacks

    private func setupWorldCallbacks() {
        world.onPieceCollision = { impulse in
            if impulse > 18 {
                Haptics.rigid()
            } else {
                Haptics.soft()
            }
        }

        world.onCushionBounce = {
            Haptics.soft()
        }

        world.onPiecePocketed = { [weak self] piece, _ in
            guard let self else { return }
            self.pocketedThisTurn.append(piece)

            if piece.type == .striker {
                Haptics.error()
            } else if piece.type == .queen {
                Haptics.boundary()
            } else {
                Haptics.success()
            }
        }
    }

    // MARK: - Board Setup

    func resetBoard() {
        pendingAction = nil
        lastFrameTime = 0
        shotPower = 0.70
        isDraggingAim = false
        aimResult = nil
        world.pieces = CarromEngine.generateStartingPieces()
        player1Score = 0
        player2Score = 0
        player1Pieces.removeAll()
        player2Pieces.removeAll()
        turn = .player1
        queenState = .onBoard
        baselineNormX = 0.5
        phase = .placement
        pocketedThisTurn.removeAll()
        banner = nil

        updateStrikerPlacement()
    }

    // MARK: - Baseline Positioning

    func updateBaselineSlider(_ value: CGFloat) {
        guard canControl else { return }
        guard phase == .placement || phase == .aiming else { return }
        baselineNormX = min(max(value, 0.0), 1.0)
        updateStrikerPlacement()
    }

    private func updateStrikerPlacement() {
        let pos = CarromEngine.baselinePosition(for: turn, normalizedX: baselineNormX)
        if let strikerIndex = world.pieces.firstIndex(where: { $0.type == .striker }) {
            world.pieces[strikerIndex].position = pos
            world.pieces[strikerIndex].velocity = .zero
            world.pieces[strikerIndex].isPocketed = false
            world.pieces[strikerIndex].sinkProgress = 0
            world.pieces[strikerIndex].pocketIndex = nil
        }
        isPlacementValid = CarromEngine.isStrikerPlacementValid(at: pos, pieces: world.pieces)

        // Default aim direction
        let defaultAimY: CGFloat = (turn == .player1) ? -1.0 : 1.0
        aimDirection = CGVector(dx: 0, dy: defaultAimY)
        updateAimGuide()
    }

    // MARK: - Aim & Strike

    func setAimPoint(_ targetPt: CGPoint) {
        guard canControl else { return }
        guard (phase == .placement || phase == .aiming), isPlacementValid else { return }
        guard let striker = world.pieces.first(where: { $0.type == .striker }) else { return }

        let toTarget = (targetPt - striker.position).vector
        if toTarget.length > 12 {
            // Check that it's aiming forward (away from baseline)
            let isForward = (turn == .player1) ? (toTarget.dy < 10) : (toTarget.dy > -10)
            if isForward {
                aimDirection = toTarget.normalized()
                if shotPower <= 0.05 {
                    shotPower = 0.70 // Default confident strike power
                }
                phase = .aiming
                updateAimGuide()
            }
        }
    }

    func setShotPower(_ power: CGFloat) {
        guard canControl else { return }
        shotPower = min(max(power, 0.1), 1.0)
        if phase == .placement && isPlacementValid {
            phase = .aiming
            updateAimGuide()
        }
    }

    func nudgeBaseline(_ delta: CGFloat) {
        guard canControl else { return }
        guard phase == .placement || phase == .aiming else { return }
        baselineNormX = min(max(baselineNormX + delta, 0.0), 1.0)
        updateStrikerPlacement()
        Haptics.soft()
    }

    func updateAimGesture(dragOffset: CGSize) {
        guard canControl else { return }
        guard phase == .placement || phase == .aiming, isPlacementValid else { return }
        phase = .aiming
        isDraggingAim = true

        // Slingshot control: dragging OPPOSITE to shoot direction
        let pullVector = CGVector(dx: -dragOffset.width, dy: -dragOffset.height)
        let pullLength = pullVector.length

        if pullLength > 10 {
            aimDirection = pullVector.normalized()
            shotPower = min(max(pullLength / 110.0, 0.20), 1.0)
        }

        updateAimGuide()
    }

    func fireStrike(power: CGFloat? = nil) {
        guard canControl else { return }
        guard (phase == .aiming || phase == .placement), isPlacementValid else { return }
        let p = power ?? (shotPower > 0.08 ? shotPower : 0.70)
        guard let strikerIndex = world.pieces.firstIndex(where: { $0.type == .striker }) else { return }

        // Calibrated carrom impulse velocities (420 to 1050 pt/s)
        let minSpeed: CGFloat = 420.0
        let maxSpeed: CGFloat = 1050.0
        let speed = minSpeed + (p * (maxSpeed - minSpeed))
        world.pieces[strikerIndex].velocity = aimDirection.normalized() * speed

        lastFrameTime = 0 // Reset time anchor so first simulation tick has clean dt
        Haptics.boundary()
        phase = .inMotion
        pocketedThisTurn.removeAll()
        aimResult = nil
        shotPower = 0
        isDraggingAim = false
    }

    func releaseStrike() {
        guard phase == .aiming || phase == .placement else { return }
        isDraggingAim = false

        guard isPlacementValid, shotPower > 0.12 else {
            phase = .placement
            shotPower = 0
            aimResult = nil
            return
        }

        guard let strikerIndex = world.pieces.firstIndex(where: { $0.type == .striker }) else { return }

        let minSpeed: CGFloat = 420.0
        let maxSpeed: CGFloat = 1050.0
        let speed = minSpeed + (shotPower * (maxSpeed - minSpeed))
        world.pieces[strikerIndex].velocity = aimDirection.normalized() * speed

        lastFrameTime = 0 // Reset time anchor
        Haptics.boundary()
        phase = .inMotion
        pocketedThisTurn.removeAll()
        aimResult = nil
        shotPower = 0
    }

    func updateAimGuide() {
        guard let striker = world.pieces.first(where: { $0.type == .striker }) else { return }
        aimResult = world.predictAim(from: striker.position, direction: aimDirection)
    }

    // MARK: - Simulation Loop Step

    func stepSimulationFrame(currentTime: TimeInterval) {
        let dt: CGFloat
        if lastFrameTime == 0 {
            dt = 1.0 / 60.0
        } else {
            let elapsed = currentTime - lastFrameTime
            dt = CGFloat(min(max(elapsed, 0.001), 0.033))
        }
        lastFrameTime = currentTime

        if let pending = pendingAction {
            if pending.delay <= Double(dt) {
                pendingAction = nil
                pending.action()
            } else {
                pendingAction = (pending.delay - Double(dt), pending.action)
            }
        }
        guard phase == .inMotion else { return }
        world.step(dt: dt)
        animationTick &+= 1

        if world.isSettled {
            lastFrameTime = 0
            evaluateTurn()
        }
    }

    // MARK: - Turn Evaluation & Rules

    private func evaluateTurn() {
        phase = .evaluating
        let ownColour: CarromPieceType = turn == .player1 ? .white : .black
        let strikerPocketed = pocketedThisTurn.contains { $0.type == .striker }
        let ownMen = pocketedThisTurn.filter { $0.type == ownColour }
        let hasQueen = pocketedThisTurn.contains { $0.type == .queen }
        var extraTurn = !ownMen.isEmpty && !strikerPocketed

        if hasQueen && !strikerPocketed {
            if !ownMen.isEmpty {
                queenState = .covered(by: turn)
                banner = CarromBanner(text: "Queen covered!", type: .success)
            } else {
                queenState = .pendingCover(by: turn)
                banner = CarromBanner(text: "Pocket your colour next to cover the queen", type: .info)
            }
            extraTurn = true
        } else if case .pendingCover(let claimant) = queenState, claimant == turn {
            if !strikerPocketed && !ownMen.isEmpty {
                queenState = .covered(by: turn)
                banner = CarromBanner(text: "Queen covered!", type: .success)
            } else {
                returnQueenToCenter()
                queenState = .onBoard
                banner = CarromBanner(text: "Cover missed — queen returned", type: .foul)
            }
        }

        if strikerPocketed {
            if hasQueen {
                returnQueenToCenter()
                queenState = .onBoard
            }
            if let penalty = world.pieces.firstIndex(where: { $0.type == ownColour && $0.isPocketed }) {
                returnPieceToCenter(at: penalty)
            }
            extraTurn = false
            banner = CarromBanner(text: "Striker foul — turn lost", type: .foul)
        }

        // The queen must be covered before either colour can clear the board.
        if case .covered = queenState {
            // A covered queen stays off the board for either player's finish.
        } else {
            for colour in [CarromPieceType.white, .black] {
                if !world.pieces.contains(where: { $0.type == colour && !$0.isPocketed }),
                   let last = world.pieces.firstIndex(where: { $0.type == colour && $0.isPocketed }) {
                    returnPieceToCenter(at: last)
                    if colour == ownColour { extraTurn = false }
                    banner = CarromBanner(text: "Cover the queen before your last piece", type: .info)
                }
            }
        }

        // Colours belong to players, including pieces accidentally pocketed by their opponent.
        player1Pieces = world.pieces.filter { $0.type == .white && $0.isPocketed }.map(\.type)
        player2Pieces = world.pieces.filter { $0.type == .black && $0.isPocketed }.map(\.type)
        player1Score = player1Pieces.count
        player2Score = player2Pieces.count
        let opponentCleared = turn == .player1 ? player2Score == targetScore : player1Score == targetScore
        let ownCleared = turn == .player1 ? player1Score == targetScore : player2Score == targetScore
        if opponentCleared || ownCleared {
            let winner = opponentCleared ? turn.opposite : turn
            phase = .gameOver(winner: winner == .player1 ? (isBot ? "You" : "Player 1") : (isBot ? "Carrom Bot" : "Player 2"))
            return
        }

        schedule(after: 1.2) { [weak self] in
            guard let self else { return }
            if !extraTurn { self.turn = self.turn.opposite }
            self.phase = .placement
            self.banner = nil
            self.shotPower = 0.70
            self.updateStrikerPlacement()
            if self.turn == .player2 && self.isBot { self.scheduleBotShot() }
        }
    }

    private func returnQueenToCenter() {
        if let idx = world.pieces.firstIndex(where: { $0.type == .queen }) {
            returnPieceToCenter(at: idx)
        }
    }

    private func returnPieceToCenter(at idx: Int) {
        let center = CGPoint(x: CarromTheme.surfaceSize / 2, y: CarromTheme.surfaceSize / 2)
        // Return near the centre without overlapping another disc.
        let otherPieces = world.pieces.filter { !$0.isPocketed && $0.id != world.pieces[idx].id }
        var returnPosition = center
        search: for ring in 0...12 {
            for slot in 0..<36 {
                let angle = CGFloat(slot) * .pi / 18
                let candidate = center + CGPoint(x: cos(angle), y: sin(angle)) * CGFloat(ring * 12)
                if otherPieces.allSatisfy({ candidate.distance(to: $0.position) >= $0.radius + CarromTheme.pieceRadius + 1 }) {
                    returnPosition = candidate
                    break search
                }
            }
        }
        world.pieces[idx].position = returnPosition
        world.pieces[idx].velocity = .zero
        world.pieces[idx].isPocketed = false
        world.pieces[idx].sinkProgress = 0
        world.pieces[idx].pocketIndex = nil
    }

    // MARK: - Bot Turn Routine

    private func scheduleBotShot() {
        guard phase == .placement, turn == .player2, isBot else { return }

        // Natural pause for realism
        schedule(after: 0.8) { [weak self] in
            guard let self, self.turn == .player2, self.phase == .placement else { return }

            let botShot = CarromEngine.calculateBotShot(pieces: self.world.pieces)

            // 1. Move slider to bot position
            withAnimation(.spring(response: 0.35, dampingFraction: 0.85)) {
                self.baselineNormX = botShot.baselineX
                self.updateStrikerPlacement()
            }

            // If no free baseline exists, end this turn instead of leaving the bot stuck.
            guard self.isPlacementValid else {
                self.banner = CarromBanner(text: "Baseline blocked — turn passed", type: .info)
                self.turn = .player1
                self.updateStrikerPlacement()
                return
            }

            // 2. Aim and draw trajectory
            self.schedule(after: 0.6) { [weak self] in
                guard let self, self.turn == .player2 else { return }
                self.aimDirection = botShot.aimDirection
                self.shotPower = min(max((botShot.power - 420.0) / 630.0, 0.20), 1.0)
                self.phase = .aiming
                self.updateAimGuide()

                // 3. Release Strike
                self.schedule(after: 0.5) { [weak self] in
                    guard let self, self.turn == .player2 else { return }
                    self.releaseStrike()
                }
            }
        }
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/games/carrom/CarromGame.kt
package com.voiid.app.main.games.carrom

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.voiid.app.ui.components.VoiidHaptics
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

sealed interface CarromPhase {
    data object Placement : CarromPhase
    data object Aiming : CarromPhase
    data object InMotion : CarromPhase
    data object Evaluating : CarromPhase
    data class GameOver(val winner: String) : CarromPhase
}

enum class CarromBannerType { INFO, SUCCESS, FOUL }
data class CarromBanner(val text: String, val type: CarromBannerType)

sealed interface QueenState {
    data object OnBoard : QueenState
    data class PendingCover(val by: CarromTurn) : QueenState
    data class Covered(val by: CarromTurn) : QueenState
}

/**
 * Port of iOS `CarromGame.swift`: the turn lifecycle, queen cover rules, fouls and the bot.
 * State is Compose state so the board and HUD recompose; the physics world is plain objects
 * stepped by [stepFrame] and read through [tick].
 */
class CarromGame(private val haptics: VoiidHaptics?, private val botDifficulty: Int = 2) {
    val isBot = true
    var turn by mutableStateOf(CarromTurn.PLAYER1); private set
    var phase by mutableStateOf<CarromPhase>(CarromPhase.Placement); private set
    var player1Score by mutableIntStateOf(0); private set
    var player2Score by mutableIntStateOf(0); private set
    var baselineNormX by mutableStateOf(0.5f); private set
    var aimDirection by mutableStateOf(Vec(0f, -1f)); private set
    var shotPower by mutableStateOf(0f); private set
    var isPlacementValid by mutableStateOf(true); private set
    var aimResult by mutableStateOf<CarromAimResult?>(null); private set
    var banner by mutableStateOf<CarromBanner?>(null); private set
    var queenState by mutableStateOf<QueenState>(QueenState.OnBoard); private set
    /** Bumped every simulated frame so the board redraws while pieces move. */
    var tick by mutableIntStateOf(0); private set

    val targetScore = 9
    val world = CarromPhysicsWorld()
    private val pocketedThisTurn = mutableListOf<CarromPiece>()
    private var lastFrameNanos = 0L
    private var pending: Pair<Float, () -> Unit>? = null

    val canControl: Boolean
        get() = (phase == CarromPhase.Placement || phase == CarromPhase.Aiming) && (turn == CarromTurn.PLAYER1 || !isBot)

    init {
        world.onPieceCollision = { impulse -> if (impulse > 18) haptics?.rigid() else haptics?.soft() }
        world.onCushionBounce = { haptics?.soft() }
        world.onPiecePocketed = { piece, _ ->
            pocketedThisTurn += piece
            when (piece.type) {
                CarromPieceType.STRIKER -> haptics?.error()
                CarromPieceType.QUEEN -> haptics?.boundary()
                else -> haptics?.success()
            }
        }
        resetBoard()
    }

    private fun schedule(delay: Float, action: () -> Unit) { pending = delay to action }
    fun pauseClock() { lastFrameNanos = 0L }

    fun resetBoard() {
        pending = null; lastFrameNanos = 0L
        shotPower = 0.70f
        aimResult = null
        world.pieces = CarromEngine.generateStartingPieces()
        player1Score = 0; player2Score = 0
        turn = CarromTurn.PLAYER1
        queenState = QueenState.OnBoard
        baselineNormX = 0.5f
        phase = CarromPhase.Placement
        pocketedThisTurn.clear()
        banner = null
        updateStrikerPlacement()
        tick++
    }

    private val striker get() = world.pieces.first { it.type == CarromPieceType.STRIKER }

    fun updateBaselineSlider(value: Float) {
        if (!canControl) return
        baselineNormX = min(max(value, 0f), 1f)
        updateStrikerPlacement()
    }

    fun nudgeBaseline(delta: Float) {
        if (!canControl) return
        baselineNormX = min(max(baselineNormX + delta, 0f), 1f)
        updateStrikerPlacement()
        haptics?.soft()
    }

    private fun updateStrikerPlacement() {
        val pos = CarromEngine.baselinePosition(turn, baselineNormX)
        striker.apply { position = pos; velocity = Vec.ZERO; isPocketed = false; sinkProgress = 0f; pocketIndex = null }
        isPlacementValid = CarromEngine.isStrikerPlacementValid(pos, world.pieces)
        aimDirection = Vec(0f, if (turn == CarromTurn.PLAYER1) -1f else 1f)
        updateAimGuide()
        tick++
    }

    fun setAimPoint(target: Vec) {
        if (!canControl || !isPlacementValid) return
        val to = target - striker.position
        if (to.length > 12f) {
            val forward = if (turn == CarromTurn.PLAYER1) to.y < 10f else to.y > -10f
            if (forward) {
                aimDirection = to.normalized()
                if (shotPower <= 0.05f) shotPower = 0.70f
                phase = CarromPhase.Aiming
                updateAimGuide()
            }
        }
    }

    fun changeShotPower(power: Float) {
        if (!canControl) return
        shotPower = min(max(power, 0.1f), 1f)
        if (phase == CarromPhase.Placement && isPlacementValid) { phase = CarromPhase.Aiming; updateAimGuide() }
    }

    /** Slingshot: dragging behind the striker aims the opposite way; distance is power. */
    fun updateAimGesture(dragX: Float, dragY: Float) {
        if (!canControl || !isPlacementValid) return
        phase = CarromPhase.Aiming
        val pull = Vec(-dragX, -dragY)
        if (pull.length > 10f) {
            aimDirection = pull.normalized()
            shotPower = min(max(pull.length / 110f, 0.20f), 1f)
        }
        updateAimGuide()
    }

    private fun launch(power: Float) {
        val speed = 420f + power * (1050f - 420f)
        striker.velocity = aimDirection.normalized() * speed
        lastFrameNanos = 0L
        haptics?.boundary()
        phase = CarromPhase.InMotion
        pocketedThisTurn.clear()
        aimResult = null
        shotPower = 0f
    }

    fun fireStrike() {
        if (!canControl || !isPlacementValid) return
        launch(if (shotPower > 0.08f) shotPower else 0.70f)
    }

    fun releaseStrike() {
        if (phase != CarromPhase.Aiming && phase != CarromPhase.Placement) return
        if (!isPlacementValid || shotPower <= 0.12f) {
            phase = CarromPhase.Placement; shotPower = 0f; aimResult = null; return
        }
        launch(shotPower)
    }

    fun updateAimGuide() { aimResult = world.predictAim(striker.position, aimDirection) }

    /** Advance one frame. Called from the screen's frame loop with the frame time. */
    fun stepFrame(frameNanos: Long) {
        val dt = if (lastFrameNanos == 0L) 1f / 60f
        else ((frameNanos - lastFrameNanos) / 1e9f).coerceIn(0.001f, 0.033f)
        lastFrameNanos = frameNanos
        pending?.let { (delay, action) ->
            if (delay <= dt) { pending = null; action() } else pending = (delay - dt) to action
        }
        if (phase != CarromPhase.InMotion) return
        world.step(dt)
        tick++
        if (world.isSettled) { lastFrameNanos = 0L; evaluateTurn() }
    }

    private fun evaluateTurn() {
        phase = CarromPhase.Evaluating
        val own = if (turn == CarromTurn.PLAYER1) CarromPieceType.WHITE else CarromPieceType.BLACK
        val strikerPocketed = pocketedThisTurn.any { it.type == CarromPieceType.STRIKER }
        val ownMen = pocketedThisTurn.filter { it.type == own }
        val hasQueen = pocketedThisTurn.any { it.type == CarromPieceType.QUEEN }
        var extraTurn = ownMen.isNotEmpty() && !strikerPocketed

        val qs = queenState
        if (hasQueen && !strikerPocketed) {
            if (ownMen.isNotEmpty()) {
                queenState = QueenState.Covered(turn)
                banner = CarromBanner("Queen covered!", CarromBannerType.SUCCESS)
            } else {
                queenState = QueenState.PendingCover(turn)
                banner = CarromBanner("Pocket your colour next to cover the queen", CarromBannerType.INFO)
            }
            extraTurn = true
        } else if (qs is QueenState.PendingCover && qs.by == turn) {
            if (!strikerPocketed && ownMen.isNotEmpty()) {
                queenState = QueenState.Covered(turn)
                banner = CarromBanner("Queen covered!", CarromBannerType.SUCCESS)
            } else {
                returnQueenToCenter()
                queenState = QueenState.OnBoard
                banner = CarromBanner("Cover missed — queen returned", CarromBannerType.FOUL)
            }
        }

        if (strikerPocketed) {
            if (hasQueen) { returnQueenToCenter(); queenState = QueenState.OnBoard }
            world.pieces.indexOfFirst { it.type == own && it.isPocketed }.takeIf { it >= 0 }?.let { returnPieceToCenter(it) }
            extraTurn = false
            banner = CarromBanner("Striker foul — turn lost", CarromBannerType.FOUL)
        }

        // The queen must be covered before either colour can clear the board.
        if (queenState !is QueenState.Covered) {
            for (colour in listOf(CarromPieceType.WHITE, CarromPieceType.BLACK)) {
                val last = world.pieces.indexOfFirst { it.type == colour && it.isPocketed }
                if (world.pieces.none { it.type == colour && !it.isPocketed } && last >= 0) {
                    returnPieceToCenter(last)
                    if (colour == own) extraTurn = false
                    banner = CarromBanner("Cover the queen before your last piece", CarromBannerType.INFO)
                }
            }
        }

        player1Score = world.pieces.count { it.type == CarromPieceType.WHITE && it.isPocketed }
        player2Score = world.pieces.count { it.type == CarromPieceType.BLACK && it.isPocketed }
        val opponentCleared = if (turn == CarromTurn.PLAYER1) player2Score == targetScore else player1Score == targetScore
        val ownCleared = if (turn == CarromTurn.PLAYER1) player1Score == targetScore else player2Score == targetScore
        if (opponentCleared || ownCleared) {
            val winner = if (opponentCleared) turn.opposite else turn
            phase = CarromPhase.GameOver(if (winner == CarromTurn.PLAYER1) "You" else "Carrom Bot")
            tick++
            return
        }

        schedule(1.2f) {
            if (!extraTurn) turn = turn.opposite
            phase = CarromPhase.Placement
            banner = null
            shotPower = 0.70f
            updateStrikerPlacement()
            if (turn == CarromTurn.PLAYER2 && isBot) scheduleBotShot()
        }
        tick++
    }

    private fun returnQueenToCenter() {
        world.pieces.indexOfFirst { it.type == CarromPieceType.QUEEN }.takeIf { it >= 0 }?.let { returnPieceToCenter(it) }
    }

    private fun returnPieceToCenter(idx: Int) {
        val c = Vec(CarromTheme.SURFACE / 2, CarromTheme.SURFACE / 2)
        val others = world.pieces.filter { !it.isPocketed && it.id != world.pieces[idx].id }
        var pos = c
        search@ for (ring in 0..12) for (slot in 0 until 36) {
            val a = (slot * PI / 18).toFloat()
            val cand = c + Vec(cos(a), sin(a)) * (ring * 12f)
            if (others.all { cand.distance(it.position) >= it.radius + CarromTheme.PIECE_RADIUS + 1f }) { pos = cand; break@search }
        }
        world.pieces[idx].apply { position = pos; velocity = Vec.ZERO; isPocketed = false; sinkProgress = 0f; pocketIndex = null }
    }

    private fun scheduleBotShot() {
        if (phase != CarromPhase.Placement || turn != CarromTurn.PLAYER2) return
        schedule(0.8f) {
            if (turn != CarromTurn.PLAYER2 || phase != CarromPhase.Placement) return@schedule
            val shot = CarromEngine.calculateBotShot(world.pieces, botDifficulty)
            baselineNormX = shot.baselineX
            updateStrikerPlacement()
            if (!isPlacementValid) {
                banner = CarromBanner("Baseline blocked — turn passed", CarromBannerType.INFO)
                turn = CarromTurn.PLAYER1
                updateStrikerPlacement()
                return@schedule
            }
            schedule(0.6f) {
                if (turn != CarromTurn.PLAYER2) return@schedule
                aimDirection = shot.aimDirection
                shotPower = min(max((shot.power - 420f) / 630f, 0.20f), 1f)
                phase = CarromPhase.Aiming
                updateAimGuide()
                schedule(0.5f) { if (turn == CarromTurn.PLAYER2) releaseStrike() }
            }
        }
    }
}
